OpenRA.Mods.Common/Traits/Buildings/Gate.cs

```diff
@@ -0,0 +1,144 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public interface ITemporaryBlocker
+	{
+		bool CanLiftBlockade(Actor self, Actor blocking);
+		bool Blocks();
+	}
+
+	[Desc("Will open and be passable for actors that appear friendly when there are no enemies in range.")]
+	public class GateInfo : BuildingInfo, ITraitInfo, Requires<WithSpriteBodyInfo>
+	{
+		public readonly string OpenSequence = "open";
+		public readonly string ClosingSequence = "closing";
+		public readonly string ClosedSequence = "closed";
+
+		public readonly string OpeningSound = "gateup1.aud";
+		public readonly string ClosingSound = "gatedwn1.aud";
+
+		[Desc("How many ticks until the gate will close after being opened.")]
+		public readonly int TicksToStayOpen = 50;
+
+		[Desc("How many ticks until the gate is open.")]
+		public readonly int TicksToOpen = 10;
+
+		public override object Create(ActorInitializer init) { return new Gate(init, this); }
+	}
+
+	public class Gate : Building, ITick, ITemporaryBlocker, INotifyBlockingMove
+	{
+		readonly GateInfo info;
+		readonly Actor self;
+		readonly WithSpriteBody wsb;
+
+		bool isOpen;
+		bool opening;
+
+		int remainingTime;
+
+		public Gate(ActorInitializer init, GateInfo info) : base(init, info)
+		{
+			this.info = info;
+			this.self = init.Self;
+
+			wsb = self.Trait<WithSpriteBody>();
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (self.IsDisabled() || !BuildComplete)
+				return;
+
+			if (opening && remainingTime-- <= 0)
+				Opened();
+			else if (isOpen)
+			{
+				if (remainingTime-- <= 0)
+				{
+					if (!IsBlocked())
+						Close();
+					else
+						remainingTime = info.TicksToStayOpen;
+				}
+				else if (remainingTime % 10 == 0 && IsBlocked())
```

Comment:
No, currently the door will only open when "nudged" from OnNotifyBlockingMove. It is always considered to be in the way, since it's invisible to the pathfinder (for units that can actually open it). There is no prediction happening, I don't remember this being done in tiberian sun either. `IsBlocked()` checks if there is an actor standing on an opened gate, which will then wait for another `info.TicksToStayOpen` to close.

Enemies can still move into your base when they sneak on an opened gate as its influenced is removed from `ActorMap` while open